peer/record.go

```diff
@@ -0,0 +1,221 @@
+// does not match the PeerID contained in the record.
+var ErrPeerIdMismatch = errors.New("signing key does not match record.PeerID")
+
+// PeerRecord contains information that is broadly useful to share with other peers,
+// either through a direct exchange (as in the libp2p identify protocol), or through
+// a Peer Routing provider, such as a DHT.
+//
+// Currently, a PeerRecord contains the public listen addresses for a peer, but this
+// is expected to expand to include other information in the future.
+//
+// PeerRecords are ordered in time by their Seq field. Newer PeerRecords must have
+// greater Seq values than older records. The NewPeerRecord function will create
+// a PeerRecord with a timestamp-based Seq value. The other PeerRecord fields should
+// be set by the caller:
+//
+//    rec := peer.NewPeerRecord()
+//    rec.PeerID = aPeerID
+//    rec.Addrs = someAddrs
+//
+// Alternatively, you can construct a PeerRecord struct directly and use the TimestampSeq
+// helper to set the Seq field:
+//
+//    rec := peer.PeerRecord{PeerID: aPeerID, Addrs: someAddrs, Seq: peer.TimestampSeq()}
+//
+// Failing to set the Seq field will not result in an error, however, a PeerRecord with a
+// Seq value of zero may be ignored or rejected by other peers.
+//
+// PeerRecords are intended to be shared with other peers inside a signed
+// routing.Envelope, and PeerRecord implements the routing.Record interface
+// to facilitate this.
+//
+// To share a PeerRecord, first call Sign to wrap the record in a Envelope
+// and sign it with the local peer's private key:
+//
+//     rec := &PeerRecord{PeerID: myPeerId, Addrs: myAddrs}
+//     envelope, err := rec.Sign(myPrivateKey)
+//
+// The resulting record.Envelope can be marshalled to a []byte and shared
+// publicly. As a convenience, the MarshalSigned method will produce the
+// Envelope and marshal it to a []byte in one go:
+//
+//     rec := &PeerRecord{PeerID: myPeerId, Addrs: myAddrs}
+//     recordBytes, err := rec.MarshalSigned(myPrivateKey)
+//
+// To validate and unmarshal a signed PeerRecord from a remote peer,
+// "consume" the containing envelope, which will return both the
+// routing.Envelope and the inner Record. The Record must be cast to
+// a PeerRecord pointer before use:
+//
+//     envelope, untypedRecord, err := ConsumeEnvelope(envelopeBytes, PeerRecordEnvelopeDomain)
+//     if err != nil {
+//       handleError(err)
+//       return
+//     }
+//     peerRec := untypedRecord.(*PeerRecord)
+//
+type PeerRecord struct {
+	// PeerID is the ID of the peer this record pertains to.
+	PeerID ID
+
+	// Addrs contains the public addresses of the peer this record pertains to.
+	Addrs []ma.Multiaddr
+
+	// Seq is a monotonically-increasing sequence counter that's used to order
+	// PeerRecords in time. The interval between Seq values is unspecified,
+	// but newer PeerRecords MUST have a greater Seq value than older records
+	// for the same peer.
+	Seq uint64
+}
+
```

Comment:
Mind adding an interface check here to assert that `PeerRecord` satisfies `record.Record`?

```go
var _ record.Record = (*PeerRecord)(nil)
```

It'll bark if we ever make a change that breaks the interface.